### PR TITLE
fix(experimental elaborator): Fix global values used in the elaborator

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves an issue where globals would not have their type set correctly in the elaborator. This is because `elaborate_let` was called which creates a fresh DefinitionId rather than the one already defined by the global, and pushes the type to that fresh id.

## Summary\*

Adds a `elaborate_global_let` function which copies over the new type to the original DefinitionId defined by the global.
This PR also adds a call to `self.type_variables.clear()` after a global is resolved so that its type is not defaulted too early.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/elaborator/expressions.rs [hunk 1]
```diff
@@ -348,6 +348,8 @@ impl<'context> Elaborator<'context> {
                 let func_type =
                     self.type_check_variable(function_name, function_id, turbofish_generics);
 
+                self.interner.push_expr_type(function_id, func_type.clone());
```
Comment (jfecher) [inline compiler/noirc_frontend/src/elaborator/expressions.rs:351]: This is an unrelated one line fix I forgot about. We weren't pushing a function identifier's type before after desugaring a method call into a function call.

File: tooling/nargo_cli/tests/stdlib-tests.rs [hunk 2]
```diff
@@ -59,15 +60,3 @@ fn run_stdlib_tests(use_elaborator: bool) {
...
-fn stdlib_noir_tests() {
-    run_stdlib_tests(false)
-}
-
-// Once this no longer panics we can use the elaborator by default and remove the old passes
-#[test]
-#[should_panic]
-fn stdlib_elaborator_tests() {
-    run_stdlib_tests(true)
-}
```
Comment (jfecher) [inline tooling/nargo_cli/tests/stdlib-tests.rs:73]: The elaborator no longer panics on any test so I had to remove this

File: compiler/noirc_frontend/src/elaborator/mod.rs [hunk 3]
```diff
@@ -1227,11 +1225,12 @@ impl<'context> Elaborator<'context> {
...
-        let (let_statement, _typ) = self.elaborate_let(let_stmt);
+        self.elaborate_global_let(let_stmt, global_id);
 
-        let statement_id = self.interner.get_global(global_id).let_statement;
-        self.interner.get_global_definition_mut(global_id).kind = definition_kind.clone();
-        self.interner.replace_statement(statement_id, let_statement);
+        // Avoid defaulting the types of globals here since they may be used in any function.
+        // Otherwise we may prematurely default to a Field inside the next function if this
+        // global was unused there, even if it is consistently used as a u8 everywhere else.
+        self.type_variables.clear();
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/elaborator/mod.rs:1233]: Could we potentially be clearing type variables for non-globals that we do want defaulted?

Comment (vezenovm) [reply compiler/noirc_frontend/src/elaborator/mod.rs:1233]: Would it be better to do something more similar to recover_generics? recover_type_variables?

**Review by vezenovm — APPROVED** [review]

File: compiler/noirc_frontend/src/elaborator/mod.rs (hunk 3 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/elaborator/mod.rs:1233]: The `type_variables` field was originally just in the TypeChecker for keeping track of type variables created in functions, it wasn't really meant to be used outside that context. We could add extra tracking so that functions only default type variables they create by remembering the old size of this field then have a check at the very end of the program for defaulting any globals afterward. That's unneeded though since monomorphization will later catch any globals and either default them or error.

Comment (vezenovm) [reply compiler/noirc_frontend/src/elaborator/mod.rs:1233]: Ok thanks for the clarification
